content/blog/2019-11-21-bug-finding/index.md

```diff
@@ -0,0 +1,308 @@
+
+### Bug Types
+
+Before we move on, I would like to introduce bug types for understanding the experiment results:
+
+1. A *crash error* is one that crashes the compiler during compilation.
+2. A *wrong error* is one that crashes the program during run-time.
+3. A *slient wrong error* is a wrong error that does not produce a compiler warning during compilation.
+4. An *assertion failure* is an LLVM internal consistency check failure.
+5. An *internal compiler failure* is a GCC internal consistency check failure.
+
+### Quantitative Comparison of GCC and LLVM Versions
+
+The following figure shows the compilation and execution results of LLVM 1.9–2.8, GCC 3.[0–4].0, and GCC 4.[0–5].0 given the input of 1,000,000 Csmith randomly generated programs. Every program was compiled at –O0, –O1, –O2, –Os, and –O3. A test case is valid if every compiler terminated within five minutes and if every compiled random program terminated within 5 seconds. The top and bottom row are the error rates of different compiler versions. The authors also find the source of the compilation error and plot it in the middle row.
+
+<img src="llvm_gcc.png" style="width: 100%">
+
+### Bug-Finding Performance as a Function of Test-Case Size
+
+The goal of designing Csmith is to find many defects quickly, and to what size the program should Csmith generate to achieve that goal becomes a question. When reporting the error, we preferred smaller programs over larger one because they are easier to debug and report. The figure below shows the experiment performed to learn the error number and runtime tradeoff given the same run-time. 
+
+*Discussion Questions:*
+1. Is there a big benefit in using small vs. large programs as tests?
+2. Most tests we write are relatively small; are there bugs that can only be caught with really large programs?
+3. How small would our program space need to be in order to "exhaustively" search for bugs?
+
+<img src="size_error.jpg" style="width: 60%">
+
+### Bug-Finding Performance Compared to Other Tools
+
+Finally, there is also a performance test of Csmith against other bug finding tools. Given the same time of bug finding, Csmith is much more efficient than existing testing tools.
+
+<img src="other-tool.jpg" style="width: 60%">
+
+### Code Coverage
+
+Finally, there is also a code coverage test on Csmith generated programs as shown in the following table. 
+
+<img src="coverage.jpg" style="width: 70%">
+
+Adding 10,000 Csmith generated program to the existing test suite of LLVM and GCC plus 10,000 Csmith does not to increase the coverage a lot. 
```

Comment:
This sentence makes two references to the 10k Csmith programs… were they added twice? Or just once?